avatica/src/main/java/org/apache/calcite/avatica/AvaticaUtils.java

```diff
@@ -200,25 +202,32 @@ public static Class box(Class clazz) {
 
   /** Reads the contents of an input stream and returns as a string. */
   public static String readFully(InputStream inputStream) throws IOException {
-    return _readFully(inputStream).toString();
+    final byte[] bytes = _readFully(inputStream);
+    return new String(bytes, 0, bytes.length, StandardCharsets.UTF_8);
   }
 
   public static byte[] readFullyToBytes(InputStream inputStream) throws IOException {
-    return _readFully(inputStream).toByteArray();
+    return _readFully(inputStream);
   }
 
-  /** Reads the contents of an input stream and returns a ByteArrayOutputStrema. */
-  static ByteArrayOutputStream _readFully(InputStream inputStream) throws IOException {
+  /**
+   * Reads the contents of an input stream and returns a byte array.
+   *
+   * @param inputStream the input to read from.
+   * @return A byte array whose length is equal to the number of bytes contained.
+   */
+  static byte[] _readFully(InputStream inputStream) throws IOException {
     final byte[] bytes = new byte[4096];
-    final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+    // Avoid the synchronization on ByteArrayOutputStream
+    final UnsynchronizedBuffer.Writer writer = new UnsynchronizedBuffer.Writer(4096);
```

Comment:
Ran a dirty test. Wrote `byte[]`s of various size into the impl here as well as a list-backed variant. There doesn't seem to be a big difference.

```
list-of-buffers
1198ms
1044ms
791ms
801ms
703ms
796ms
759ms
826ms
834ms
781ms

Current
1048ms
852ms
666ms
664ms
660ms
676ms
664ms
696ms
672ms
708ms
```
We can investigate this one outside of these changes if you'd like. I think the current approach is a nice improvement.